In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

In [2]:
df = pd.read_csv(r'C:\Users\laptop\Downloads\cc2dca92b57d11ec\dataset\train.csv')
df.head()

ID        date warehouse_ID   Latitude   Longitude Product_Type  year  \
0  0x2710  2017-01-01     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
1  0x33e6  2017-01-01     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
2  0x2711  2017-01-02     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
3  0x33e7  2017-01-02     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
4  0x2712  2017-01-03     WH_0x3e9  41.681471  -72.794746       Type_A  2017   

     month is_weekend is_warehouse_closed  daily_dispatch_count  \
0  January        Yes                  No                   5.5   
1  January        Yes                  No                   6.1   
2  January         No                  No                   4.8   
3  January         No                  No                   6.7   
4  January         No                  No                   5.8   

   weekly_dispatch_count  
0                    5.5  
1                    6.1  
2                    NaN  
3                    NaN  
4                    NaN

In [3]:
df.info()    #Checking the datetime format of the date column 

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16644 entries, 0 to 16643
Data columns (total 12 columns):
ID                       16644 non-null object
date                     16644 non-null object
warehouse_ID             16644 non-null object
Latitude                 16312 non-null float64
Longitude                16228 non-null float64
Product_Type             16644 non-null object
year                     16644 non-null int64
month                    16644 non-null object
is_weekend               16145 non-null object
is_warehouse_closed      16478 non-null object
daily_dispatch_count     16644 non-null float64
weekly_dispatch_count    2372 non-null float64
dtypes: float64(4), int64(1), object(7)
memory usage: 1.5+ MB


In [4]:
df['date'] = pd.to_datetime(df['date'])

In [5]:
df.isnull().sum()

ID                           0
date                         0
warehouse_ID                 0
Latitude                   332
Longitude                  416
Product_Type                 0
year                         0
month                        0
is_weekend                 499
is_warehouse_closed        166
daily_dispatch_count         0
weekly_dispatch_count    14272
dtype: int64

In [6]:
len(df)

16644

In [7]:
df['day'] = df['date'].dt.day     # Adding a day column, an extra information

In [8]:
df['month'] = df['date'].dt.month   # Changing the month as strings to numbers

In [9]:
df.head()

ID       date warehouse_ID   Latitude   Longitude Product_Type  year  \
0  0x2710 2017-01-01     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
1  0x33e6 2017-01-01     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
2  0x2711 2017-01-02     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
3  0x33e7 2017-01-02     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
4  0x2712 2017-01-03     WH_0x3e9  41.681471  -72.794746       Type_A  2017   

   month is_weekend is_warehouse_closed  daily_dispatch_count  \
0      1        Yes                  No                   5.5   
1      1        Yes                  No                   6.1   
2      1         No                  No                   4.8   
3      1         No                  No                   6.7   
4      1         No                  No                   5.8   

   weekly_dispatch_count  day  
0                    5.5    1  
1                    6.1    1  
2                    NaN    2  
3                    NaN    2  
4                    NaN    3

In [10]:
df.Latitude.unique()   

array([41.681471, 38.749077,       nan, 35.67897 , 34.361664, 25.750156,
       30.189375, 41.611603, 48.692551, 37.915825, 61.600803])

In [11]:
df.Longitude.unique()

array([ -72.794746, -105.18306 ,         nan, -109.067413,  -86.305595,
        -80.27964 ,  -82.648254,  -87.551186, -116.319862,  -89.829262,
       -149.125259])

Thus, we can see that the dataset have only 10 pairs of (Latitude, Longitude). So, we don't need any Regression model to fit them. We, will group them assign the NaNs to the value according to the pairs. If we find both pairs to be null, we will perform a diiferent statergy to fill them.

In [12]:
df['Latitude'] = df['Latitude'].fillna(0)
df['Longitude'] = df['Longitude'].fillna(0)

In [13]:
 # All the unique Latitude and Longitude values are listed here
    
la = [41.681471, 38.749077, 35.67897 , 34.361664, 25.750156, 30.189375,
       41.611603, 37.915825, 48.692551, 61.600803]                 
lo = [-72.794746, -105.18306 , -109.067413,  -86.305595,  -80.27964 ,
        -82.648254,  -87.551186,  -89.829262, -116.319862, -149.125259]

In [14]:
L1 = list(df['Latitude'])
L1 = [round(x,6) for x in L1]
L2 = list(df['Longitude'])
L2 = [round(x,6) for x in L2]

In [15]:
for i,j in zip(range(len(L1)), range(len(L2))):
    for k,l in zip(range(len(la)),range(len(lo))):
        if L1[i] == 0 and L2[j] == lo[l]:
            L1[i] = la[k]
        elif L1[i] == la[k] and L2[j] == 0:
            L2[j] = lo[l]

In [16]:
df = df.reset_index(drop=True)
df['Latitude'] = pd.DataFrame(L1)
df['Longitude'] = pd.DataFrame(L2)

In [17]:
df.head()

ID       date warehouse_ID   Latitude   Longitude Product_Type  year  \
0  0x2710 2017-01-01     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
1  0x33e6 2017-01-01     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
2  0x2711 2017-01-02     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
3  0x33e7 2017-01-02     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
4  0x2712 2017-01-03     WH_0x3e9  41.681471  -72.794746       Type_A  2017   

   month is_weekend is_warehouse_closed  daily_dispatch_count  \
0      1        Yes                  No                   5.5   
1      1        Yes                  No                   6.1   
2      1         No                  No                   4.8   
3      1         No                  No                   6.7   
4      1         No                  No                   5.8   

   weekly_dispatch_count  day  
0                    5.5    1  
1                    6.1    1  
2                    NaN    2  
3                    NaN    2  
4                    NaN    3

In [18]:
df.isnull().sum()

ID                           0
date                         0
warehouse_ID                 0
Latitude                     0
Longitude                    0
Product_Type                 0
year                         0
month                        0
is_weekend                 499
is_warehouse_closed        166
daily_dispatch_count         0
weekly_dispatch_count    14272
day                          0
dtype: int64

In [19]:
df['is_weekend'] = df['is_weekend'].fillna(0)

In [20]:
g1 = list(df['is_weekend'])

In [21]:
for i in range(len(g1)):
    if i%2 == 0 and g1[i] == 0:
        g1[i] = g1[i+1]
    elif i%2 != 0 and g1[i] == 0:
        g1[i] = g1[i-1]

In [22]:
for i in range(len(g1)):
    if g1[i] == 0.0:
        print(i)
        
# 0's actaully mean that, its unknown if it's a weekend or not, for all the observations for that day. so will filter them out.                         

370
371
7320
7321
7690
7691
8794
8795
10542
10543
12712
12713
14874
14875
15222
15223


In [23]:
df['is_weekend'] = pd.DataFrame(g1)

In [24]:
df.isnull().sum()

ID                           0
date                         0
warehouse_ID                 0
Latitude                     0
Longitude                    0
Product_Type                 0
year                         0
month                        0
is_weekend                   0
is_warehouse_closed        166
daily_dispatch_count         0
weekly_dispatch_count    14272
day                          0
dtype: int64

In [25]:
len(df[df['is_warehouse_closed']=='No'])

15365

In [26]:
len(df[df['is_warehouse_closed']=='Yes'])

1113

So, more than 93% of the known days the 'Warehouse' is not closed. Thus, we can safely assume that most the days (166 days = 1%) in which data is unavaliable, the 'Warehouse' is not closed.

In [32]:
df['is_warehouse_closed'] = df['is_warehouse_closed'].apply(lambda x: 'No' if pd.isnull(x) else x)

In [34]:
df.isnull().sum()

ID                           0
date                         0
warehouse_ID                 0
Latitude                     0
Longitude                    0
Product_Type                 0
year                         0
month                        0
is_weekend                   0
is_warehouse_closed          0
daily_dispatch_count         0
weekly_dispatch_count    14272
day                          0
dtype: int64

In [62]:
df.to_csv('raw_data.csv')

In [58]:
df = df[df['Latitude'] != 0.0]

In [60]:
df = df[df['is_weekend'] !=0.0]

In [61]:
len(df)

16616

# Handelling the missing values from the target Dataset
At first, we have to handle the 14272 missing datasets out of the 16644, from the target value. There are several common ways of handelling the missing datasets listed as:

(i)   Removimg the rows with missing datasets, which is not possible here because we will lose about 86% of the data.

(ii)  Using statistical measures like Mean, Median & Mode to fill the missing values. The mode way been used previously, but in Regression it might not help.

(iii) Selecting a Machine Learning model to predict the missing values. This will be the best choice for a highly accurate model and let's see if we can use it. 

(iv) Using a Clustering technique, but this is not applicable here, because it can't used for a 'Regression problem'.

In [63]:
test = df[df['weekly_dispatch_count'].isnull()]

In [64]:
train = df.dropna(subset=['weekly_dispatch_count'])

In [65]:
len(train)

2370

In [66]:
train.head()

ID       date warehouse_ID   Latitude   Longitude Product_Type  year  \
0   0x2710 2017-01-01     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
1   0x33e6 2017-01-01     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
14  0x2717 2017-01-08     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
15  0x33ed 2017-01-08     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
28  0x271e 2017-01-15     WH_0x3e9  41.681471  -72.794746       Type_A  2017   

    month is_weekend is_warehouse_closed  daily_dispatch_count  \
0       1        Yes                  No                   5.5   
1       1        Yes                  No                   6.1   
14      1        Yes                  No                   4.5   
15      1        Yes                  No                   6.5   
28      1        Yes                  No                   4.6   

    weekly_dispatch_count  day  
0                     5.5    1  
1                     6.1    1  
14                   37.1    8  
15                   48.3    8  
28                   37.5   15

In [68]:
Train1 = train.drop(['weekly_dispatch_count','ID','date'],axis=1)

In [69]:
Train1.isnull().sum()

warehouse_ID            0
Latitude                0
Longitude               0
Product_Type            0
year                    0
month                   0
is_weekend              0
is_warehouse_closed     0
daily_dispatch_count    0
day                     0
dtype: int64

# Dummy variables classification
 
  All the Dummy variables Warehouse_id, is_weekend, is_warehouse_closed and Product_Type are of 'Nominal' type, so we have to use 'get_dummies' feature of pandas.
            
            
           

In [70]:
Train1 = pd.get_dummies(Train1,columns=['warehouse_ID', 'is_weekend', 'is_warehouse_closed', 'Product_Type'])

In [72]:
np.shape(train)

(2370, 13)

In [73]:
y = train['weekly_dispatch_count']

In [202]:
# As, we have to forcast the 'daily_dispatch_count' also, that means this also a dependent variable. Thus, we need to drop it.
x = Train1.drop('daily_dispatch_count',axis=1)

In [203]:
x.head()

Latitude   Longitude  year  month  day  warehouse_ID_WH_0x3e9  \
0   41.681471  -72.794746  2017      1    1                      1   
1   38.749077 -105.183060  2017      1    1                      0   
14  41.681471  -72.794746  2017      1    8                      1   
15  38.749077 -105.183060  2017      1    8                      0   
28  41.681471  -72.794746  2017      1   15                      1   

    warehouse_ID_WH_0x3ea  warehouse_ID_WH_0x3eb  warehouse_ID_WH_0x3ec  \
0                       0                      0                      0   
1                       1                      0                      0   
14                      0                      0                      0   
15                      1                      0                      0   
28                      0                      0                      0   

    warehouse_ID_WH_0x3ed  warehouse_ID_WH_0x3ee  warehouse_ID_WH_0x3ef  \
0                       0                      0                      0   
1                       0                      0                      0   
14                      0                      0                      0   
15                      0                      0                      0   
28                      0                      0                      0   

    warehouse_ID_WH_0x3f0  warehouse_ID_WH_0x3f1  warehouse_ID_WH_0x3f2  \
0                       0                      0                      0   
1                       0                      0                      0   
14                      0                      0                      0   
15                      0                      0                      0   
28                      0                      0                      0   

    is_weekend_Yes  is_warehouse_closed_No  is_warehouse_closed_Yes  \
0                1                       1                        0   
1                1                       1                        0   
14               1                       1                        0   
15               1                       1                        0   
28               1                       1                        0   

    Product_Type_Type_A  Product_Type_Type_B  
0                     1                    0  
1                     1                    0  
14                    1                    0  
15                    1                    0  
28                    1                    0

# Choosing the most effective Machine Learning Regression model
Here, as we have a mixture of the Continous, Binary and Categorical Features, it will good to use a 'Descision Tree Regressor' or a'Random Forest Regressor' for fitting the training and testing data to deploy the model. 

In [204]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split

In [205]:
x_train, x_test, y_train, y_test = train_test_split(x,y, test_size=0.2)

In [227]:
Dreg = DecisionTreeRegressor(max_depth=7)
Dreg.fit(x_train, y_train)

DecisionTreeRegressor(max_depth=7)

In [238]:
Dreg.score(x_test,y_test)

0.9308230738983577

In [239]:
Dreg.score(x_train, y_train)

0.9488644132037247

In [224]:
rreg = RandomForestRegressor(n_estimators=1000, max_depth=7)
rreg.fit(x_train,y_train)

RandomForestRegressor(max_depth=7, n_estimators=1000)

In [240]:
rreg.score(x_test,y_test)

0.941496668444141

In [241]:
rreg.score(x_train, y_train)

0.9591437728823201

The (x_test, y_test) score is very high for the 'Random Forest Regressor' and, also there is not much overfitting present in 
the model. But, on the grounds of accuracy score we observ that, the 'Random Forest Regressor' performs better than the 'Descision Tree Regressor'.  This is why, we will use the 'Random Forest Regressor' model for predicting the
 missing values for the 'weekly_dispatch_count'.

# Handelling the created 'test' data 

In [91]:
test.isnull().sum()         #Checking the NaN values 

ID                           0
date                         0
warehouse_ID                 0
Latitude                     0
Longitude                    0
Product_Type                 0
year                         0
month                        0
is_weekend                   0
is_warehouse_closed          0
daily_dispatch_count         0
weekly_dispatch_count    14246
day                          0
dtype: int64

In [92]:
test.drop('weekly_dispatch_count',axis = 1, inplace=True)

C:\Users\laptop\Anaconda31\lib\site-packages\pandas\core\frame.py:3940: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  errors=errors)


In [93]:
test.head()

ID       date warehouse_ID   Latitude   Longitude Product_Type  year  \
2  0x2711 2017-01-02     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
3  0x33e7 2017-01-02     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
4  0x2712 2017-01-03     WH_0x3e9  41.681471  -72.794746       Type_A  2017   
5  0x33e8 2017-01-03     WH_0x3ea  38.749077 -105.183060       Type_A  2017   
6  0x2713 2017-01-04     WH_0x3e9  41.681471  -72.794746       Type_A  2017   

   month is_weekend is_warehouse_closed  daily_dispatch_count  day  
2      1         No                  No                   4.8    2  
3      1         No                  No                   6.7    2  
4      1         No                  No                   5.8    3  
5      1         No                  No                   8.1    3  
6      1         No                  No                   6.6    4

In [94]:
test1 = pd.get_dummies(test, columns=['warehouse_ID', 'is_weekend', 'is_warehouse_closed', 'Product_Type'])

In [95]:
test1 = test1.drop(['date','ID'],axis=1)

In [230]:
test1 = test1.drop('daily_dispatch_count',axis=1)

In [231]:
test1.head()

Latitude   Longitude  year  month  day  warehouse_ID_WH_0x3e9  \
2  41.681471  -72.794746  2017      1    2                      1   
3  38.749077 -105.183060  2017      1    2                      0   
4  41.681471  -72.794746  2017      1    3                      1   
5  38.749077 -105.183060  2017      1    3                      0   
6  41.681471  -72.794746  2017      1    4                      1   

   warehouse_ID_WH_0x3ea  warehouse_ID_WH_0x3eb  warehouse_ID_WH_0x3ec  \
2                      0                      0                      0   
3                      1                      0                      0   
4                      0                      0                      0   
5                      1                      0                      0   
6                      0                      0                      0   

   warehouse_ID_WH_0x3ed  warehouse_ID_WH_0x3ee  warehouse_ID_WH_0x3ef  \
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   
5                      0                      0                      0   
6                      0                      0                      0   

   warehouse_ID_WH_0x3f0  warehouse_ID_WH_0x3f1  warehouse_ID_WH_0x3f2  \
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   
5                      0                      0                      0   
6                      0                      0                      0   

   is_weekend_Yes  is_warehouse_closed_No  is_warehouse_closed_Yes  \
2               0                       1                        0   
3               0                       1                        0   
4               0                       1                        0   
5               0                       1                        0   
6               0                       1                        0   

   Product_Type_Type_A  Product_Type_Type_B  
2                    1                    0  
3                    1                    0  
4                    1                    0  
5                    1                    0  
6                    1                    0

In [232]:
test1.columns

Index(['Latitude', 'Longitude', 'year', 'month', 'day',
       'warehouse_ID_WH_0x3e9', 'warehouse_ID_WH_0x3ea',
       'warehouse_ID_WH_0x3eb', 'warehouse_ID_WH_0x3ec',
       'warehouse_ID_WH_0x3ed', 'warehouse_ID_WH_0x3ee',
       'warehouse_ID_WH_0x3ef', 'warehouse_ID_WH_0x3f0',
       'warehouse_ID_WH_0x3f1', 'warehouse_ID_WH_0x3f2', 'is_weekend_Yes',
       'is_warehouse_closed_No', 'is_warehouse_closed_Yes',
       'Product_Type_Type_A', 'Product_Type_Type_B'],
      dtype='object')

In [233]:
test1.drop(['is_weekend_No'],axis=1,inplace=True)
Train1.columns

Index(['Latitude', 'Longitude', 'year', 'month', 'daily_dispatch_count', 'day',
       'warehouse_ID_WH_0x3e9', 'warehouse_ID_WH_0x3ea',
       'warehouse_ID_WH_0x3eb', 'warehouse_ID_WH_0x3ec',
       'warehouse_ID_WH_0x3ed', 'warehouse_ID_WH_0x3ee',
       'warehouse_ID_WH_0x3ef', 'warehouse_ID_WH_0x3f0',
       'warehouse_ID_WH_0x3f1', 'warehouse_ID_WH_0x3f2', 'is_weekend_Yes',
       'is_warehouse_closed_No', 'is_warehouse_closed_Yes',
       'Product_Type_Type_A', 'Product_Type_Type_B'],
      dtype='object')

Unfortunately, in the 'train' data the avaliable 'weekly_dispatch_count' is only for the weekends, so will we have to remove 
an extra column from the 'test1'. This will make the columns identical for both 'Train1' and 'test1'.


In [235]:
test1.isnull().sum()              #Checking if there are still any NaN values.

Latitude                   0
Longitude                  0
year                       0
month                      0
day                        0
warehouse_ID_WH_0x3e9      0
warehouse_ID_WH_0x3ea      0
warehouse_ID_WH_0x3eb      0
warehouse_ID_WH_0x3ec      0
warehouse_ID_WH_0x3ed      0
warehouse_ID_WH_0x3ee      0
warehouse_ID_WH_0x3ef      0
warehouse_ID_WH_0x3f0      0
warehouse_ID_WH_0x3f1      0
warehouse_ID_WH_0x3f2      0
is_weekend_Yes             0
is_warehouse_closed_No     0
is_warehouse_closed_Yes    0
Product_Type_Type_A        0
Product_Type_Type_B        0
dtype: int64

In [101]:
test_v = test1.copy()

In [242]:
weekly = rreg.predict(test1)
weekly

array([18.02917926, 22.520098  , 36.69789834, ..., 10.40076044,
       36.08640785, 10.10563867])

In [243]:
Dreg.predict(test1)

array([ 5.5       ,  6.1       , 33.75      , ..., 10.48888889,
       31.36      , 10.48888889])

From the above values, we can conclude that the predicted output for both the values are not very different. But we will 
choose the best model here.

In [244]:
f1 =pd.DataFrame(weekly)

In [245]:
f1.rename(columns={0:'weekly_dispatch_count'}, inplace=True)

In [246]:
test_v.columns

Index(['Latitude', 'Longitude', 'year', 'month', 'daily_dispatch_count', 'day',
       'warehouse_ID_WH_0x3e9', 'warehouse_ID_WH_0x3ea',
       'warehouse_ID_WH_0x3eb', 'warehouse_ID_WH_0x3ec',
       'warehouse_ID_WH_0x3ed', 'warehouse_ID_WH_0x3ee',
       'warehouse_ID_WH_0x3ef', 'warehouse_ID_WH_0x3f0',
       'warehouse_ID_WH_0x3f1', 'warehouse_ID_WH_0x3f2', 'is_weekend_Yes',
       'is_warehouse_closed_No', 'is_warehouse_closed_Yes',
       'Product_Type_Type_A', 'Product_Type_Type_B'],
      dtype='object')

In [247]:
Train1.columns

Index(['Latitude', 'Longitude', 'year', 'month', 'daily_dispatch_count', 'day',
       'warehouse_ID_WH_0x3e9', 'warehouse_ID_WH_0x3ea',
       'warehouse_ID_WH_0x3eb', 'warehouse_ID_WH_0x3ec',
       'warehouse_ID_WH_0x3ed', 'warehouse_ID_WH_0x3ee',
       'warehouse_ID_WH_0x3ef', 'warehouse_ID_WH_0x3f0',
       'warehouse_ID_WH_0x3f1', 'warehouse_ID_WH_0x3f2', 'is_weekend_Yes',
       'is_warehouse_closed_No', 'is_warehouse_closed_Yes',
       'Product_Type_Type_A', 'Product_Type_Type_B'],
      dtype='object')

In [248]:
train_p = pd.concat([Train1,test_v], sort=False, ignore_index=True)

In [249]:
f2 = pd.DataFrame(y)

In [250]:
train_pf = pd.concat([f2,f1], ignore_index=True)

In [251]:
f2.head()

weekly_dispatch_count
0                     5.5
1                     6.1
14                   37.1
15                   48.3
28                   37.5

In [252]:
len(train_pf)

16616

In [253]:
len(train_p)

16616

In [254]:
train_f = pd.concat([train_p,train_pf],axis=1)

In [255]:
train_f.to_csv('train_f.csv')                          #This file will be used for the 2nd half of the project 

In [256]:
len(train_f)

16616

So, we have sucessfully reconstructed the input 'train' data, with by filling most of the values and dropping only around 1% of the 'NaN' values. 

#   PART 2

In [257]:
d = pd.read_csv(r'C:\Users\laptop\train_f.csv')
d.head()

Unnamed: 0   Latitude   Longitude  year  month  daily_dispatch_count  day  \
0           0  41.681471  -72.794746  2017      1                   5.5    1   
1           1  38.749077 -105.183060  2017      1                   6.1    1   
2           2  41.681471  -72.794746  2017      1                   4.5    8   
3           3  38.749077 -105.183060  2017      1                   6.5    8   
4           4  41.681471  -72.794746  2017      1                   4.6   15   

   warehouse_ID_WH_0x3e9  warehouse_ID_WH_0x3ea  warehouse_ID_WH_0x3eb  ...  \
0                      1                      0                      0  ...   
1                      0                      1                      0  ...   
2                      1                      0                      0  ...   
3                      0                      1                      0  ...   
4                      1                      0                      0  ...   

   warehouse_ID_WH_0x3ef  warehouse_ID_WH_0x3f0  warehouse_ID_WH_0x3f1  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   warehouse_ID_WH_0x3f2  is_weekend_Yes  is_warehouse_closed_No  \
0                      0               1                       1   
1                      0               1                       1   
2                      0               1                       1   
3                      0               1                       1   
4                      0               1                       1   

   is_warehouse_closed_Yes  Product_Type_Type_A  Product_Type_Type_B  \
0                        0                    1                    0   
1                        0                    1                    0   
2                        0                    1                    0   
3                        0                    1                    0   
4                        0                    1                    0   

   weekly_dispatch_count  
0                    5.5  
1                    6.1  
2                   37.1  
3                   48.3  
4                   37.5  

[5 rows x 23 columns]

In [258]:
d1 = d.drop('Unnamed: 0',axis=1)

In [259]:
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import train_test_split
from sklearn.multioutput import MultiOutputRegressor

In [260]:
X = d1.drop(['daily_dispatch_count','weekly_dispatch_count'],axis=1)
Y = d1[['daily_dispatch_count','weekly_dispatch_count']]

In [261]:
X_train, X_test, Y_train,Y_test = train_test_split(X,Y,test_size=0.2)

In [262]:
len(X_train)

13292

In [263]:
len(X_test)

3324

In [280]:
Dreg1 = DecisionTreeRegressor(max_depth=9)
Dreg1.fit(X_train, Y_train)

DecisionTreeRegressor(max_depth=9)

In [281]:
Dreg1.score(X_test,Y_test)

0.9009609506814049

In [282]:
Dreg1.score(X_train,Y_train)

0.9108284972547076

In [331]:
rreg1 = RandomForestRegressor(n_estimators = 500, max_depth=9)
rreg1.fit(X_train,Y_train)

RandomForestRegressor(max_depth=9, n_estimators=500)

In [332]:
rreg1.score(X_test,Y_test)

0.9050941581074322

In [333]:
rreg1.score(X_train,Y_train)

0.9144524545155854

   # Handelling 'Test' Dataset

In [297]:
d2 = pd.read_csv(r'C:\Users\laptop\Downloads\cc2dca92b57d11ec\dataset\test.csv')
d2.head()

ID        date warehouse_ID   Latitude   Longitude Product_Type  year  \
0  0x31fc  2021-05-01     WH_0x3e9  41.681471         NaN       Type_A  2021   
1  0x31fd  2021-05-01     WH_0x3e9  41.681471         NaN       Type_B  2021   
2  0x3ed2  2021-05-01     WH_0x3ea  38.749077 -105.183060       Type_A  2021   
3  0x3ed3  2021-05-01     WH_0x3ea  38.749077 -105.183060       Type_B  2021   
4  0x4a3a  2021-05-01     WH_0x3eb  35.678970 -109.067413       Type_A  2021   

  month is_weekend is_warehouse_closed  
0   May        NaN                  No  
1   May        Yes                  No  
2   May        Yes                  No  
3   May        Yes                  No  
4   May        Yes                 Yes

In [298]:
len(d2)

4900

In [299]:
d2.isnull().sum()

ID                       0
date                     0
warehouse_ID             0
Latitude                98
Longitude              122
Product_Type             0
year                     0
month                    0
is_weekend             147
is_warehouse_closed     49
dtype: int64

In [300]:
d3 = pd.read_csv(r'C:\Users\laptop\Downloads\cc2dca92b57d11ec\dataset\submission_weekly.csv')
d3.head()

ID  weekly_dispatch_count
0  0x31fe                    NaN
1  0x31ff                    NaN
2  0x3ed4                    NaN
3  0x3ed5                    NaN
4  0x4a3c                    NaN

In [301]:
np.shape(d3)

(700, 2)

In [302]:
d4 = d2[d2['ID'].isin(d3['ID'])]  # For selecting the rows for which we have to forcast

In [303]:
d4.head()

ID        date warehouse_ID   Latitude   Longitude Product_Type  year  \
20  0x31fe  2021-05-02     WH_0x3e9  41.681471  -72.794746       Type_A  2021   
21  0x31ff  2021-05-02     WH_0x3e9  41.681471  -72.794746       Type_B  2021   
22  0x3ed4  2021-05-02     WH_0x3ea  38.749077 -105.183060       Type_A  2021   
23  0x3ed5  2021-05-02     WH_0x3ea  38.749077 -105.183060       Type_B  2021   
24  0x4a3c  2021-05-02     WH_0x3eb  35.678970 -109.067413       Type_A  2021   

   month is_weekend is_warehouse_closed  
20   May        Yes                  No  
21   May        Yes                  No  
22   May        Yes                  No  
23   May        Yes                  No  
24   May        Yes                  No

In [304]:
d4['date'] = pd.to_datetime(d4['date'])
d4['month'] = d4['date'].dt.month
d4['day'] = d4['date'].dt.day
d4 = d4.drop('date',axis=1)

C:\Users\laptop\Anaconda31\lib\site-packages\ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  """Entry point for launching an IPython kernel.
C:\Users\laptop\Anaconda31\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy
  
C:\Users\laptop\Anaconda31\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: http://p

In [305]:
d4.head()

ID warehouse_ID   Latitude   Longitude Product_Type  year  month  \
20  0x31fe     WH_0x3e9  41.681471  -72.794746       Type_A  2021      5   
21  0x31ff     WH_0x3e9  41.681471  -72.794746       Type_B  2021      5   
22  0x3ed4     WH_0x3ea  38.749077 -105.183060       Type_A  2021      5   
23  0x3ed5     WH_0x3ea  38.749077 -105.183060       Type_B  2021      5   
24  0x4a3c     WH_0x3eb  35.678970 -109.067413       Type_A  2021      5   

   is_weekend is_warehouse_closed  day  
20        Yes                  No    2  
21        Yes                  No    2  
22        Yes                  No    2  
23        Yes                  No    2  
24        Yes                  No    2

In [306]:
d4.isnull().sum()

ID                      0
warehouse_ID            0
Latitude               13
Longitude              11
Product_Type            0
year                    0
month                   0
is_weekend             20
is_warehouse_closed     7
day                     0
dtype: int64

In [307]:
d4.to_csv('check111.csv')          #Cheking the values in excel

In [308]:
d4.Latitude.unique()

array([41.681471, 38.749077, 35.67897 , 34.361664, 25.750156, 30.189375,
       41.611603, 37.915825, 48.692551, 61.600803,       nan])

In [309]:
d4.Longitude.unique()

array([ -72.794746, -105.18306 , -109.067413,  -86.305595,  -80.27964 ,
        -82.648254,  -87.551186,  -89.829262, -116.319862, -149.125259,
               nan])

In [310]:
la = [41.681471, 38.749077, 35.67897 , 34.361664, 25.750156, 30.189375,
       41.611603, 37.915825, 48.692551, 61.600803]
lo = [-72.794746, -105.18306 , -109.067413,  -86.305595,  -80.27964 ,
        -82.648254,  -87.551186,  -89.829262, -116.319862, -149.125259]

In [311]:
d4['Latitude'] = d4['Latitude'].fillna(0)
d4['Longitude'] = d4['Longitude'].fillna(0)

In [312]:
l1 = list(d4['Latitude'])
l1 = [round(x,6) for x in l1]
l2 = list(d4['Longitude'])
l2 = [round(x,6) for x in l2]


In [313]:
for i,j in zip(range(len(l1)), range(len(l2))):
    for k,l in zip(range(len(la)),range(len(lo))):
        if l1[i] == 0 and l2[j] == lo[l]:
            l1[i] = la[k]
        elif l1[i] == la[k] and l2[j] == 0:
            l2[j] = lo[l]

In [314]:
d4['is_weekend'] = 'Yes'

In [315]:
d4 = d4.reset_index(drop=True)
d4['Latitude'] = pd.DataFrame(l1)
d4['Longitude'] = pd.DataFrame(l2)


In [316]:
d4.head()

ID warehouse_ID   Latitude   Longitude Product_Type  year  month  \
0  0x31fe     WH_0x3e9  41.681471  -72.794746       Type_A  2021      5   
1  0x31ff     WH_0x3e9  41.681471  -72.794746       Type_B  2021      5   
2  0x3ed4     WH_0x3ea  38.749077 -105.183060       Type_A  2021      5   
3  0x3ed5     WH_0x3ea  38.749077 -105.183060       Type_B  2021      5   
4  0x4a3c     WH_0x3eb  35.678970 -109.067413       Type_A  2021      5   

  is_weekend is_warehouse_closed  day  
0        Yes                  No    2  
1        Yes                  No    2  
2        Yes                  No    2  
3        Yes                  No    2  
4        Yes                  No    2

In [317]:
d4.isnull().sum()

ID                     0
warehouse_ID           0
Latitude               0
Longitude              0
Product_Type           0
year                   0
month                  0
is_weekend             0
is_warehouse_closed    7
day                    0
dtype: int64

In [319]:
d4.to_csv('new_data.csv')

In [320]:
len(d4[d4['is_warehouse_closed'] == 'Yes'])

160

In [321]:
len(d4[d4['is_warehouse_closed'] == 'No'])

533

Here, again about 77% of the cases are for the 'Warehouse' not closed in the weekend. This is why, we will set the 7 unknown 
values to 'No', which is about 1% of the data. This is the 'Mode' filling NaN method that we have already used before.  

In [322]:
d4['is_warehouse_closed'] = d4['is_warehouse_closed'].apply(lambda x: 'No' if pd.isnull(x) else x)

In [323]:
d4.isnull().sum()

ID                     0
warehouse_ID           0
Latitude               0
Longitude              0
Product_Type           0
year                   0
month                  0
is_weekend             0
is_warehouse_closed    0
day                    0
dtype: int64

In [325]:
d5 = d4.copy()
d5 = pd.get_dummies(d5,columns=['warehouse_ID', 'is_weekend', 'is_warehouse_closed', 'Product_Type'])

In [326]:
d5.head()

ID   Latitude   Longitude  year  month  day  warehouse_ID_WH_0x3e9  \
0  0x31fe  41.681471  -72.794746  2021      5    2                      1   
1  0x31ff  41.681471  -72.794746  2021      5    2                      1   
2  0x3ed4  38.749077 -105.183060  2021      5    2                      0   
3  0x3ed5  38.749077 -105.183060  2021      5    2                      0   
4  0x4a3c  35.678970 -109.067413  2021      5    2                      0   

   warehouse_ID_WH_0x3ea  warehouse_ID_WH_0x3eb  warehouse_ID_WH_0x3ec  ...  \
0                      0                      0                      0  ...   
1                      0                      0                      0  ...   
2                      1                      0                      0  ...   
3                      1                      0                      0  ...   
4                      0                      1                      0  ...   

   warehouse_ID_WH_0x3ee  warehouse_ID_WH_0x3ef  warehouse_ID_WH_0x3f0  \
0                      0                      0                      0   
1                      0                      0                      0   
2                      0                      0                      0   
3                      0                      0                      0   
4                      0                      0                      0   

   warehouse_ID_WH_0x3f1  warehouse_ID_WH_0x3f2  is_weekend_Yes  \
0                      0                      0               1   
1                      0                      0               1   
2                      0                      0               1   
3                      0                      0               1   
4                      0                      0               1   

   is_warehouse_closed_No  is_warehouse_closed_Yes  Product_Type_Type_A  \
0                       1                        0                    1   
1                       1                        0                    0   
2                       1                        0                    1   
3                       1                        0                    0   
4                       1                        0                    1   

   Product_Type_Type_B  
0                    0  
1                    1  
2                    0  
3                    1  
4                    0  

[5 rows x 21 columns]

In [327]:
d5 = d5.drop('ID',axis=1)

In [334]:
forecast = rreg1.predict(d5)

In [335]:
forecast

array([[ 4.14124279, 24.20595371],
       [ 1.93296863,  9.12075999],
       [ 6.47218442, 42.47365237],
       ...,
       [ 0.04686155, 11.29254965],
       [ 5.78855466, 40.49834774],
       [ 0.05013933, 11.12162886]])

In [339]:
d6 = pd.DataFrame(forecast)
d6.rename(columns={0:'daily_dispatch_count', 1: 'weekly_dispatch_count'},inplace=True)

In [341]:
d7 = pd.concat([d4,d6],axis=1)

In [353]:
d8 = d7[['ID','weekly_dispatch_count']]

In [354]:
d8.to_csv('submission.csv')